### ISSUE TITLE
v.4.x: AssertionFailedException in Decompiler.IL.ControlFlow.LoopDetection.MoveBlocksIntoContainer

### ISSUE THREAD

**Issue by greenozon** [icsharpcode/ILSpy#1361]
ILSpy version 4.0.0.4472-beta3

```
Error decompiling ---.WorkbooksOpen
 ---> ICSharpCode.ILSpy.AssertionFailedException: 
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message, String detailMessage) in offset 380
   at ICSharpCode.ILSpy.ILSpyTraceListener.Fail(String message) in offset 9
   at System.Diagnostics.TraceInternal.Fail(String message) in offset 55
   at System.Diagnostics.Debug.Assert(Boolean condition) in offset 6
   at ICSharpCode.Decompiler.IL.ControlFlow.LoopDetection.MoveBlocksIntoContainer(List`1 loop, BlockContainer loopContainer) in offset 156
   at ICSharpCode.Decompiler.IL.ControlFlow.LoopDetection.ConstructLoop(List`1 loop, ControlFlowNode exitPoint) in offset 165
   at ICSharpCode.Decompiler.IL.ControlFlow.LoopDetection.Run(Block block, BlockTransformContext context) in offset 579
   at ICSharpCode.Decompiler.IL.Block.RunTransforms(IEnumerable`1 transforms, BlockTransformContext context) in offset 69
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 146
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.VisitBlock(ControlFlowNode cfgNode, BlockTransformContext context) in offset 89
   at ICSharpCode.Decompiler.IL.Transforms.BlockILTransform.Run(ILFunction function, ILTransformContext context) in offset 146
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 581
-- continuing with outer exception (ICSharpCode.Decompiler.DecompilerException) --
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(IMethod method, EntityDeclaration entityDecl, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 1068
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(IMethod method, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 249
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(ITypeDefinition typeDef, DecompileRun decompileRun, ITypeResolveContext decompilationContext) in offset 747
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompileTypes(IEnumerable`1 types, DecompileRun decompileRun, ITypeResolveContext decompilationContext, SyntaxTree syntaxTree) in offset 219
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileTypes(IEnumerable`1 types) in offset 201
   at ICSharpCode.Decompiler.CSharp.WholeProjectDecompiler.<>c__DisplayClass30_0.<WriteCodeFilesInProject>b__2(IGrouping`2 file) in offset 68
   at System.Threading.Tasks.Parallel.<>c__DisplayClass17_0`1.<ForWorker>b__1() in offset 589
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask) in offset 6
   at System.Threading.Tasks.Task.<>c__DisplayClass176_0.<ExecuteSelfReplicating>b__0(Object) in offset 134

```

target: https://www.sendspace.com/file/cfsmyq

**Comment by siegfriedpammer** [comment]
Sorry, I forgot to download the binary. Could you please upload it again? Thank you very much!

**Comment by greenozon** [comment]
can't find as well, closing

**Comment by dgrunwald** [comment]
The second assembly from #1352 has this assertion in the `WorkbooksOpen` function, so I guess we found it.

**Comment by dgrunwald** [comment]
The function here uses VB `On Error Resume Next`.
The exceptional back-edge is not supported (#1092 = won't fix); but we should in theory be able to decompile the non-exceptional code into a mess of gotos without crashing.

**Comment by dgrunwald** [comment]
The control flow graph of this function looks rather interesting:
![WorkbooksOpen](https://user-images.githubusercontent.com/243140/61911842-ee25e100-af38-11e9-90f4-1ca4d14561d8.png)

`IL_0614` is the target of the unsupported exceptional back-edge, and is thus unreachable from ILSpy's point of view.
`IL_0000` is the normal entry point of the function.
Black edges are control flow, green edges are dominance.

**Comment by dgrunwald** [comment]
Basically, there's a huge switch that can jump to any block, and is unreachable because back edges from a catch block into middle of the corresponding try block are unsupported.

LoopDetection doesn't see these incoming edges because it's dominance based (unreachable code is invisible to the analysis). But that makes it impossible to form the loops, because there's no valid place to put the unreachable blocks.
Turning on the "Remove dead and side effect free code" option avoids this assertion, because the unreachable code is removed because it can cause problems for loop detection.

So I think ultimately this assertion is just another instance of #1242.

**Comment by dgrunwald** [comment]
Fixed in 8117dfff4a2b238a0e2794416c78b9d3656dbfb8.
